fendermint/vm/interpreter/src/fvm/examples/mycustomkernel.rs

```diff
@@ -0,0 +1,201 @@
+// Copyright 2021-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+use fvm::call_manager::CallManager;
+use fvm::gas::Gas;
+use fvm::kernel::filecoin::{DefaultFilecoinKernel, FilecoinKernel};
+use fvm::kernel::prelude::*;
+use fvm::kernel::Result;
+use fvm::kernel::{
+    ActorOps, CryptoOps, DebugOps, EventOps, IpldBlockOps, MessageOps, NetworkOps, RandomnessOps,
+    SelfOps, SendOps, SyscallHandler, UpgradeOps,
+};
+use fvm::syscalls::Linker;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::consensus::ConsensusFault;
+use fvm_shared::piece::PieceInfo;
+use fvm_shared::randomness::RANDOMNESS_LENGTH;
+use fvm_shared::sector::{
+    AggregateSealVerifyProofAndInfos, RegisteredSealProof, ReplicaUpdateInfo, SealVerifyInfo,
+};
+use fvm_shared::sys::out::network::NetworkContext;
+use fvm_shared::sys::out::vm::MessageContext;
+use fvm_shared::{address::Address, econ::TokenAmount, ActorID, MethodNum};
+
+use ambassador::Delegate;
+use cid::Cid;
+
+// we define a single custom syscall which simply doubles the input
+pub trait CustomKernel: Kernel {
```

Comment:
Am I right that if one wants to register a syscall only for an actor, then it's not necessary to implement a new kernel trait and it'd be enough to implement just a new kernel impl (which is called here `DefaultCustomKernel`) and corresponding `SyscallHandler` for it.

Reply:
You don't register a syscall for an individual actor, but for a whole kernel which then exposes that syscall to all actors.

If you want to register a _new_ syscall, then you have to define a new kernel trait with a new kernel implementation and corresponding SyscallHandler. 

However, If you want to override an existing syscall then you don't _need_ a new trait but just a separate implementation/corresponding SyscallHandler.

Reply:
> If you want to register a new syscall, then you have to define a new kernel trait with a new kernel implementation and corresponding SyscallHandler.

I tried to understand why, following the source code of Fendermint and came up with only idea that it's really needed if one want to change some types like `Limiter`.

Could you please explain which drawbacks are in [this](https://github.com/consensus-shipyard/ipc/compare/main...fluencelabs:ipc:fluence#diff-2f5413c018d995ea1371b7c3a906f41bfe06fdd65077103969efd1a8d5857a1cR33) approach where I define a new kernel impl and then impl `SyscallHandler` with a new syscall for it?

Reply:
Hmm, maybe I am misunderstanding but you need to somehow invoke the new syscall on the new kernel, I don't see the implementation of the `fluence_fvm_syscall::run_randomx`, where is the actual kernel implementation of the syscall? 

See similar to the [link_syscall function](https://github.com/filecoin-project/ref-fvm/blob/40d396edcd486ae5185f649ab59d52e8815975a3/testing/integration/src/custom_kernel/mod.rs#L200) which calls the new syscall on the kernel [here](https://github.com/filecoin-project/ref-fvm/blob/40d396edcd486ae5185f649ab59d52e8815975a3/testing/integration/src/custom_kernel/mod.rs#L53)

Reply:
Right, it's being invoked in directly this [function](https://github.com/fluencelabs/fluence-fendermint/blob/main/fluence-fendermint-syscall/src/lib.rs#L27). I don't see why do we need to call it from a kernel...

Reply:
Yeah, if you can implement the functionality outside of the kernel, than that should be fine!